plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/ExchangeStorageWriter.java

```diff
@@ -0,0 +1,29 @@
+package io.trino.plugin.exchange;
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import io.airlift.slice.Slice;
+
+import java.io.IOException;
+
+public interface ExchangeStorageWriter
+        extends AutoCloseable
+{
+    void write(Slice slice)
+            throws IOException;
```

Comment:
I was thinking more about the `isBlocked` problem. 

Currently on the `ExchangeSink` interface level the `isBlocked` method indicates whether the `add(partition, data)` method can be called without blocking.

Since today the size of `data` is not limited it is not possible to provide such a guarantee even if all the buffers are completely empty. Thus it feels like the first step we should do is to limit the maximum size of `data`. For direct exchange the maximum size of a single page is effectively limited by 16MB: https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/ExchangeClientConfig.java#L34 (`getMaxContentLength` is set to 16MB internally). It should be reasonable to set the maximum allowed size of `data` to 16MB.

Now let's assume there is a limit (e.g.: 16MB). It doesn't solve the problem on it's own just yet. Currently the total allocated buffer space for a single partition is only 10MB. Even if buffers for a partition are completely empty it is still not possible to guarantee that the `add` method won't block. 

To solve this issue we could change the buffering strategy a little. Instead of having two 5MB buffers attached to a specific partition we can have a common buffer pool shared between all partitions.

Here is an example of such algorithm:

1. On construction: Allocate a shared buffer pool (can be further optimized to allocate buffers lazily). Each buffer will be of a size required by the storage (5MB for S3). The total number of buffers should be twice the number of output partitions to facilitate double buffering. e.g.:  `private final Queue<Slice> freeBuffers = new ArrayDeque(numberOfPartitions * 2)`
2. On `add(partition, data)`:
    - Take a free buffer from the `freeBuffers` pool, append `data` to that buffer. 
    - If buffer is full - place it in a "flushing" queue
    - If buffer is not full - mark it as "used" and bind it to the partition
    - Next time when a new data received for a `partition` check if there's a "used" buffer assigned, if not - repeat.
3. On `isBlocked`: If the total size of all buffers in the `freeBuffers` pool is larger than `16MB` - the sink is unblocked (as there's enough buffer space to accept any `add(...)` call), otherwise blocked. 
4. The flushing has to happen asynchronously and concurrently. When a buffer is flushed it has to be placed back into the `freeBuffers` queue. 

This algorithm could also be more flexible when it comes to handling skew. Since the buffer space is shared the process is no longer restricted by the 10MB of buffer space allocated for a single partition. As long as there's enough "flushers" ("open files") the algorithm will handle the skew pretty well. 

What do you think?

Reply:
Looks cool, when it's blocked, what kind of futures should we return?

Reply:
It could be a `SettableFuture` (or a CompletableFuture analogue)